hoss/coordinate_utilities.py

```diff
@@ -0,0 +1,232 @@
+
+    # if the override is the variable
+    elif (
+        varinfo.get_variable(variable_name).is_latitude()
+        or varinfo.get_variable(variable_name).is_longitude()
+    ):
+        override_dimensions = get_override_projected_dimension_names(
+            varinfo, variable_name
+        )
+    return override_dimensions
+
+
+def get_variables_with_anonymous_dims(
+    varinfo: VarInfoFromDmr, variables: set[str]
+) -> set[str]:
+    """
+    returns a set of variables without any
+    dimensions
+    """
+    return set(
+        variable
+        for variable in variables
+        if len(varinfo.get_variable(variable).dimensions) == 0
+    )
+
+
+def get_coordinate_variables(
+    varinfo: VarInfoFromDmr,
+    requested_variables: Set[str],
+) -> tuple[list, list]:
+    """This method returns coordinate variables that are referenced
+    in the variables requested. It returns it in a specific order
+    [latitude, longitude]
+    """
+
+    coordinate_variables_set = sorted(
+        varinfo.get_references_for_attribute(requested_variables, 'coordinates')
+    )
+
+    latitude_coordinate_variables = [
+        coordinate
+        for coordinate in coordinate_variables_set
+        if varinfo.get_variable(coordinate).is_latitude()
+    ]
+
+    longitude_coordinate_variables = [
+        coordinate
+        for coordinate in coordinate_variables_set
+        if varinfo.get_variable(coordinate).is_longitude()
+    ]
+
+    return latitude_coordinate_variables, longitude_coordinate_variables
+
+
+def get_row_col_sizes_from_coordinate_datasets(
+    lat_arr: ndarray,
+    lon_arr: ndarray,
+) -> Tuple[int, int]:
+    """
+    This method returns the row and column sizes of the coordinate datasets
+
+    """
+    if lat_arr.ndim > 1 and lon_arr.shape == lat_arr.shape:
+        col_size = lat_arr.shape[1]
+        row_size = lat_arr.shape[0]
+    elif (
+        lat_arr.ndim == 1
+        and lon_arr.ndim == 1
+        and lat_arr.size > 0
+        and lon_arr.size > 0
+    ):
+        col_size = lon_arr.size
+        row_size = lat_arr.size
+    else:
+        raise IrregularCoordinateVariables(lon_arr.shape, lat_arr.shape)
+    return row_size, col_size
+
+
+def get_lat_lon_arrays(
+    prefetch_dataset: Dataset,
+    latitude_coordinate: VariableFromDmr,
+    longitude_coordinate: VariableFromDmr,
+) -> Tuple[ndarray, ndarray]:
+    """
+    This method is used to return the lat lon arrays from a 2D
+    coordinate dataset.
+    """
+    try:
+        lat_arr = prefetch_dataset[latitude_coordinate.full_name_path][:]
+    except Exception as exception:
+        raise MissingCoordinateVariable(
+            latitude_coordinate.full_name_path
+        ) from exception
+
+    try:
+        lon_arr = prefetch_dataset[longitude_coordinate.full_name_path][:]
+    except Exception as exception:
+        raise MissingCoordinateVariable(
+            longitude_coordinate.full_name_path
+        ) from exception
+
+    return lat_arr, lon_arr
```

Comment:
if the name does not match what is in the prefetch dataset it will throw an exception

Reply:
Okay. Just playing around with that a tiny bit, I think you are trying to catch an `IndexError`:

```
>>> from netCDF4 import Dataset
>>> ds = Dataset('SMAP_L3_FT_P_20240729_R19240_001.h5', 'r')
>>> ds['/not_a_variable']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "src/netCDF4/_netCDF4.pyx", line 2522, in netCDF4._netCDF4.Dataset.__getitem__
IndexError: not_a_variable not found in /
```

So that would mean you'd have: `except IndexError as exception:`.

Reply:
> I think that the near-exact duplication in this function tells you that you should just have a single function, maybe something like `get_coordinate_array`, where you just pass in the one variable, and get one array out.
> 
> The only things different in the first `try`/`except` versus the second is `latitude_coordinate.full_name_path` versus `longitude_coordinate.full_name_path`, and the name of the variable you are assigning the output to.
> 
> It would be DRYer to just do:
> 
> ```
> def get_coordinate_array(
>     prefetch_dataset: Dataset,
>     coordinate_name: str,
> ) -> np.ndarray:
>     """This function returns the `numpy` array from a
>     coordinate dataset.
> 
>     """
>     try:
>         coordinate_array = prefetch_dataset[coordinate_name][:]
>     except Exception as exception:
>         raise MissingCoordinateVariable(
>             coordinate_name
>         ) from exception`
> 
>     return coordinate_array
> ```
> 
> Then when calling the function, you'd do something like:
> 
> ```
> latitude_array = get_coordinate_array(prefetch_dataset, latitude_coordinate.full_name_path)
> longitude_array = get_coordinate_array(prefetch_dataset, longitude_coordinate.full_name_path)
> ```

updated the function - ca881d1b0eb7b365990b42581c27f2be095eecdc